File: files/CH-204.R
library(tidyverse)
library(readxl)

path = "files/CH-204 Custom Grouping.xlsx"
input = read_excel(path, range = "B2:C19")
test  = read_excel(path, range = "H2:J19")

result = input %>%
  mutate(Group = accumulate(Cost, 
                              ~ if (.x$sum + .y > 150 | .x$count == 3) 
                                list(sum = .y, count = 1, group = .x$group + 1) 
                              else 
                                list(sum = .x$sum + .y, count = .x$count + 1, group = .x$group),
                              .init = list(sum = 0, count = 0, group = 1)
  )[-1] %>% map_int("group"))

# Solution has differences in Cost so Groups are not matching.

File: files/200-299/204/CH-204.py
import pandas as pd
from itertools import accumulate

path = "CH-204 Custom Grouping.xlsx"
input = pd.read_excel(path, usecols="B:C", skiprows=1, nrows=17)
test = pd.read_excel(path, usecols="H:J", skiprows=1, nrows=17)

def accumulate_groups(costs):
    groups, acc_sum, acc_count, group = [], 0, 0, 1
    for cost in costs:
        if acc_sum + cost > 150 or acc_count == 3:
            group += 1
            acc_sum, acc_count = cost, 1
        else:
            acc_sum += cost
            acc_count += 1
        groups.append(group)
    return groups

input['Group'] = accumulate_groups(input['Cost'])

print(input)